json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php

```diff
@@ -213,24 +224,153 @@ protected function result() {
 class Jetpack_JSON_API_Sync_Now_Endpoint extends Jetpack_JSON_API_Sync_Endpoint {
 	protected function result() {
 		$args = $this->input();
+		$queue_name = $this->validate_queue( $args['queue'] );
 
-		if ( ! isset( $args['queue'] ) ) {
-			return new WP_Error( 'invalid_queue', 'Queue name is required', 400 );
+		if ( is_wp_error( $queue_name ) ){
+			return $queue_name;
 		}
 
-		if ( ! in_array( $args['queue'], array( 'sync', 'full_sync' ) ) ) {
-			return new WP_Error( 'invalid_queue', 'Queue name should be sync or full_sync', 400 );
+		require_once JETPACK__PLUGIN_DIR . 'sync/class.jetpack-sync-sender.php';
+
+		$sender = Jetpack_Sync_Sender::get_instance();
+		$response = $sender->do_sync_for_queue( new Jetpack_Sync_Queue( $queue_name ) );
+
+		return array(
+			'response' => $response
+		);
+	}
+}
+
+class Jetpack_JSON_API_Sync_Checkout_Endpoint extends Jetpack_JSON_API_Sync_Endpoint {
+	protected function result() {
+		$args = $this->query_args();
+		$queue_name = $this->validate_queue( $args['queue'] );
+
+		if ( is_wp_error( $queue_name ) ){
+			return $queue_name;
 		}
 
-		$queue_name = $args['queue'];
+		if ( is_int( $args[ 'number_of_items' ] ) && (int) $args[ 'number_of_items' ] < 1 && (int) $args[ 'number_of_items' ] > 100 ) {
+			return new WP_Error( 'invalid_number_of_items', 'Number of items needs to be an integer that is larger than 0 and less then 100', 400 );
+		}
 
 		require_once JETPACK__PLUGIN_DIR . 'sync/class.jetpack-sync-sender.php';
+		require_once JETPACK__PLUGIN_DIR . 'sync/class.jetpack-sync-queue.php';
+		require_once JETPACK__PLUGIN_DIR . 'sync/class.jetpack-sync-json-deflate-array-codec.php';
+
+		$queue = new Jetpack_Sync_Queue( $queue_name );
+		$codec = new Jetpack_Sync_JSON_Deflate_Array_Codec();
 
+		// We need the sender to set up all the before send module actions...
 		$sender = Jetpack_Sync_Sender::get_instance();
-		$response = $sender->do_sync_for_queue( new Jetpack_Sync_Queue( $queue_name ) );
+		
+		$skipped_items_ids = array();
+
+		// let's delete the checkin state
+		if ( isset( $args['force'] ) && $args['force'] ) {
+			$queue->force_checkin();
+		}
+
+		$encode = ( $args['encode'] ? true : false );
+
+		$codec_name = $encode ? $codec->name() : null;
+
+		$buffer = $this->get_buffer( $queue, $args[ 'number_of_items' ] );
+		
+		// Check that the $buffer is not checkout out already
+		if ( is_wp_error( $buffer ) ) {
+			return new WP_Error( 'buffer_open', "We couldn't get the buffer it is currently checked out", 400 );
+		}
+		
+		if ( ! is_object( $buffer ) ) {
+			return new WP_Error( 'buffer_non-object', 'Buffer is not an object', 400 );
+		}
 
+		$items = $buffer->get_items();
+
+		// set up current screen to avoid errors rendering content
```

Comment:
it would be good to move this logic to its own method. Could be done in another PR :)